File: app/src/main/java/com/example/sl_terms/DataBaseTest.java
package com.example.sl_terms;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class DataBaseTest {
    private static final String GET_AVAILABLE_TEST = "http://btidb.ru/available_tests.php";
    private static final String GET_ID_QUESTIONS = "http://btidb.ru/id_questions.php?id_test=";
    private static final String GET_TEXT_QUESTION = "http://btidb.ru/name_question.php?id_question=";
    private static final String INSERT_STUDENT = "http://btidb.ru/Insert_student.php?surname=";
    private static final String GET_ID_VARIANT_NAME_VARIANT = "http://btidb.ru/id_variant_and_variant_name.php?id_question=";
    private static final String ANSWER_TO_CUR_QUESTION = "http://btidb.ru/insert_answer.php?id_student=";
    private static final String NUMBER_OF_CORRECT_ANSWERS = "http://btidb.ru/number_of_correct_answers.php?id_student=";
    private static final String PICTURE = "http://btidb.ru/picture.php?id_question=";
    private static final String PICTURE_NULL = "http://btidb.ru/null_picture.php?id_question=";


    private final OkHttpClient client = new OkHttpClient();

    static public DataBaseTest dbt;

    String getResponse(String url) {
        String strResponse = "";
        Request request = new Request.Builder().url(url).build();
        try {
            Response response = client.newCall(request).execute();
            strResponse = response.body().string();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return strResponse;
    }
    //получить список доступных тестов
    AvailableTest[] getAvailableTest() {
        ArrayList<AvailableTest> listTest = new ArrayList<>();
        try {
            JSONArray dataJsonArr = new JSONObject(getResponse(GET_AVAILABLE_TEST))
                    .getJSONArray("tests");
            for (int i = 0; i < dataJsonArr.length(); i++) {
                JSONObject testJSON = dataJsonArr.getJSONObject(i);
                AvailableTest availabletest = new AvailableTest();
                availabletest.id = testJSON.getInt("id");
                availabletest.name = testJSON.getString("name");

                System.out.println(availabletest.id);
                System.out.println(availabletest.name);
                System.out.println("11111111111");
                listTest.add(availabletest);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return listTest.toArray(new AvailableTest[listTest.size()]);
    }

    //вернуть все вопросы данного теста (айди)
    AvailableTest[] getIdQuestion(int id_test) {
        ArrayList<AvailableTest> listTest = new ArrayList<>();
        try {
            JSONArray dataJsonArr = new JSONObject(getResponse(GET_ID_QUESTIONS + id_test))
                    .getJSONArray("id_questions");
            for (int i = 0; i < dataJsonArr.length(); i++) {
                JSONObject testJSON = dataJsonArr.getJSONObject(i);
                AvailableTest availabletest = new AvailableTest();

                availabletest.id = testJSON.getInt("id");

                listTest.add(availabletest);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return listTest.toArray(new AvailableTest[listTest.size()]);
    }

    //получить текст вопроса по айди
    AvailableTest[] getTextQuestion(int id_question) {
        ArrayList<AvailableTest> listTest = new ArrayList<>();
        try {
            JSONArray dataJsonArr = new JSONObject(getResponse(GET_TEXT_QUESTION + id_question))
                    .getJSONArray("name_questions");
            for (int i = 0; i < dataJsonArr.length(); i++) {
                JSONObject testJSON = dataJsonArr.getJSONObject(i);
                AvailableTest availabletest = new AvailableTest();

                availabletest.name = testJSON.getString("text");

                listTest.add(availabletest);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return listTest.toArray(new AvailableTest[listTest.size()]);
    }


    //добавить студента в бд
    int insertStudent(String FIO) {
        int id_student = 0;
        ArrayList<AvailableTest> listTest = new ArrayList<>();
        try {
            JSONArray dataJsonArr = new JSONObject(getResponse(INSERT_STUDENT + FIO))
                    .getJSONArray("id_student");
            for (int i = 0; i < dataJsonArr.length(); i++) {
                JSONObject testJSON = dataJsonArr.getJSONObject(i);
                AvailableTest availabletest = new AvailableTest();

                availabletest.id = testJSON.getInt("id_student");
               // System.out.println("2222222222");
               // System.out.println(availabletest.id);
                listTest.add(availabletest);

                if(i == (dataJsonArr.length()-1)){
                    id_student = availabletest.id;
                }
            }

        } catch (Exception e) {
            e.printStackTrace();
        }

        System.out.println(id_student);
        return id_student;
    }


    //получить варианты ответов на конкретный вопрос
    AvailableTest[] geIdVariantVariantName(int id_question) {
        ArrayList<AvailableTest> listTest = new ArrayList<>();
        try {
            JSONArray dataJsonArr = new JSONObject(getResponse(GET_ID_VARIANT_NAME_VARIANT + id_question))
                    .getJSONArray("variants");
            for (int i = 0; i < dataJsonArr.length(); i++) {
                JSONObject testJSON = dataJsonArr.getJSONObject(i);
                AvailableTest availabletest = new AvailableTest();
                availabletest.id = testJSON.getInt("id_variant");
                availabletest.name = testJSON.getString("variant_name");
                System.out.println(availabletest.id);
                System.out.println(availabletest.name);
                listTest.add(availabletest);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return listTest.toArray(new AvailableTest[listTest.size()]);
    }
    //отправить вариант ответа
    void answerToCurQuestion(int id_student, int id_question, int id_variant, int id_session){
        getResponse(ANSWER_TO_CUR_QUESTION + id_student + "&id_question=" + id_question + "&id_answer="
                + id_variant + "&id_session=" + id_session);
    }

    //

    //получить количество правильных ответов
    int numberOfCorrectAnswers(int id_student) {
        int number = 0;
        ArrayList<AvailableTest> listTest = new ArrayList<>();
        try {
            JSONArray dataJsonArr = new JSONObject(getResponse(NUMBER_OF_CORRECT_ANSWERS + id_student))
                    .getJSONArray("count");
            for (int i = 0; i < dataJsonArr.length(); i++) {
                JSONObject testJSON = dataJsonArr.getJSONObject(i);
                AvailableTest availabletest = new AvailableTest();

                availabletest.id = testJSON.getInt("count");

                listTest.add(availabletest);

                if(i == (dataJsonArr.length()-1)){
                    number = availabletest.id;
                }
            }

        } catch (Exception e) {
            e.printStackTrace();
        }

        return number;
    }
    AvailableTest[] getPicture(int id_question) {
        ArrayList<AvailableTest> listTest = new ArrayList<>();
        try {
            JSONArray dataJsonArr = new JSONObject(getResponse(PICTURE + id_question))
                    .getJSONArray("variants");
            for (int i = 0; i < dataJsonArr.length(); i++) {
                JSONObject testJSON = dataJsonArr.getJSONObject(i);
                AvailableTest availabletest = new AvailableTest();
                availabletest.id = testJSON.getInt("id_picture");
                availabletest.name = testJSON.getString("picture");
                System.out.println(availabletest.id);
                System.out.println(availabletest.name);
                listTest.add(availabletest);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return listTest.toArray(new AvailableTest[listTest.size()]);
    }
    AvailableTest[] getPictureNull(int id_question) {
        ArrayList<AvailableTest> listTest = new ArrayList<>();
        try {
            JSONArray dataJsonArr = new JSONObject(getResponse(PICTURE_NULL + id_question))
                    .getJSONArray("variants");
            for (int i = 0; i < dataJsonArr.length(); i++) {
                JSONObject testJSON = dataJsonArr.getJSONObject(i);
                AvailableTest availabletest = new AvailableTest();
                availabletest.id = testJSON.getInt("id_picture");
                System.out.println(availabletest.id);
                listTest.add(availabletest);
            }
        } catch (Exception e) {
            e.printStackTrace();
            AvailableTest availabletest = new AvailableTest();
            availabletest.id = 0;
            listTest.add(availabletest);
            return listTest.toArray(new AvailableTest[listTest.size()]);
        }
        return listTest.toArray(new AvailableTest[listTest.size()]);
    }
    boolean checkInternet() {
        return !getResponse(GET_AVAILABLE_TEST ).equals("");
    }
}

File: app/src/main/java/com/example/sl_terms/DataBaseTest.kt
package com.example.sl_terms

import com.example.sl_terms.models.AvailableTest
import com.example.sl_terms.models.Option
import com.example.sl_terms.models.Question
import com.squareup.moshi.JsonAdapter
import com.squareup.moshi.Moshi
import com.squareup.moshi.Types
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject
import java.lang.reflect.Type
import java.util.*

class DataBaseTest {
    private val client = OkHttpClient()
    fun getResponse(url: String): String {
        var strResponse = ""
        val request = Request.Builder().url(url).build()
        try {
            val response = client.newCall(request).execute()
            strResponse = response.body()?.string().toString()
            println("$url -- ${response.code()}")
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return strResponse
    }

    fun getResponseObj(url: String): Any {
        data class Response (val code: Int, val body: String)
        var responseCode: Int = 200
        var responseStr: String = ""
        val request = Request.Builder().url(url).build()
        try {
            val response = client.newCall(request).execute()
            responseCode = response.code()
            if (response.body() != null) {
                responseStr = response.body().toString()
            }

        } catch (e: Exception) {
            e.printStackTrace()
        }
        return Response(code = responseCode, body = responseStr)
    }

    //получить список доступных тестов
    val availableTest: Array<AvailableTest>
        get() {
            val listTest = ArrayList<AvailableTest>()
            try {
                val dataJsonArr = JSONObject(getResponse(GET_AVAILABLE_TEST))
                        .getJSONArray("tests")
                for (i in 0 until dataJsonArr.length()) {
                    val testJSON = dataJsonArr.getJSONObject(i)
                    println(testJSON)
                    val availabletest = AvailableTest()
                    availabletest.id = testJSON.getInt("id")
                    availabletest.name = testJSON.getString("name")
                    println(availabletest.id)
                    println(availabletest.name)
                    println("11111111111")
                    listTest.add(availabletest)
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
            // Test data
//            val availabletest = AvailableTest()
//            availabletest.id = 0
//            availabletest.name = "Test test"
//            listTest.add(AvailableTest())
            return listTest.toTypedArray()
        }

    //вернуть все вопросы данного теста (айди)
    fun getIdQuestion(id_test: Int): Array<AvailableTest> {
        val listTest = ArrayList<AvailableTest>()
        try {
            val dataJsonArr = JSONObject(getResponse(GET_ID_QUESTIONS + id_test))
                    .getJSONArray("id_questions")
            for (i in 0 until dataJsonArr.length()) {
                val testJSON = dataJsonArr.getJSONObject(i)
                val availabletest = AvailableTest()
                availabletest.id = testJSON.getInt("id")
                listTest.add(availabletest)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return listTest.toTypedArray()
    }

    //получить текст вопроса по айди
    fun getTextQuestion(id_question: Int): Array<AvailableTest> {
        val listTest = ArrayList<AvailableTest>()
        try {
            val dataJsonArr = JSONObject(getResponse(GET_TEXT_QUESTION + id_question))
                    .getJSONArray("name_questions")
            for (i in 0 until dataJsonArr.length()) {
                val testJSON = dataJsonArr.getJSONObject(i)
                val availabletest = AvailableTest()
                availabletest.name = testJSON.getString("text")
                listTest.add(availabletest)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return listTest.toTypedArray()
    }

    fun getQuestions(id_test: Int): Array<Question> {
        var listQuestions: List<Question> = listOf()
        try {
            val listType: Type = Types.newParameterizedType(List::class.java, Question::class.java)
            val dataJson = getResponse(GET_ID_QUESTIONS + id_test)
            val moshi = Moshi.Builder()
                    .add(KotlinJsonAdapterFactory())
                    .build()
            val questionAdapter: JsonAdapter<List<Question>> = moshi.adapter(listType)
            listQuestions = questionAdapter.fromJson(dataJson)!!
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return listQuestions.toTypedArray()
    }

    //добавить студента в бд
    fun insertStudent(FIO: String): Int {
        var id_student = 0
        try {
            val dataJson = JSONObject(getResponse(INSERT_STUDENT + FIO))
            id_student = dataJson.getInt("idStudent")
        } catch (e: Exception) {
            e.printStackTrace()
        }
        println(id_student)
        return id_student
    }

    //получить варианты ответов на конкретный вопрос
    fun geIdVariantVariantName(id_question: Int): Array<AvailableTest> {
        val listTest = ArrayList<AvailableTest>()
        try {
            val dataJsonArr = JSONObject(getResponse(GET_ID_VARIANT_NAME_VARIANT + id_question))
                    .getJSONArray("variants")
            for (i in 0 until dataJsonArr.length()) {
                val testJSON = dataJsonArr.getJSONObject(i)
                val availabletest = AvailableTest()
                availabletest.id = testJSON.getInt("id_variant")
                availabletest.name = testJSON.getString("variant_name")
                println(availabletest.id)
                println(availabletest.name)
                listTest.add(availabletest)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return listTest.toTypedArray()
    }

    fun getOptions(id_question: Int): Array<Option> {
        var listOptions: List<Option> = listOf()
        try {
            val listType: Type = Types.newParameterizedType(List::class.java, Option::class.java)
            val dataJson = getResponse(GET_ID_VARIANT_NAME_VARIANT + id_question)
            val moshi = Moshi.Builder()
                    .add(KotlinJsonAdapterFactory())
                    .build()
            val optionAdapter: JsonAdapter<List<Option>> = moshi.adapter(listType)
            listOptions = optionAdapter.fromJson(dataJson)!!
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return listOptions.toTypedArray()
    }

    //отправить вариант ответа
    fun answerToCurQuestion(id_student: Int, id_question: Int, id_variant: Int, id_session: Int) {
        getResponse(ANSWER_TO_CUR_QUESTION + id_student + "&id_question=" + id_question + "&id_answer="
                + id_variant + "&id_session=" + id_session)
    }

    //
//получить количество правильных ответов
    fun numberOfCorrectAnswers(id_student: Int, idTest: Int): Int {
        var number = 0
        try {
            val dataJson = JSONObject(getResponse("$NUMBER_OF_CORRECT_ANSWERS$id_student&id_test=$idTest"))
            number = dataJson.getInt("count")
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return number
    }

    fun getPicture(id_question: Int): Array<AvailableTest> {
        val listTest = ArrayList<AvailableTest>()
        try {
            val dataJsonArr = JSONObject(getResponse(PICTURE + id_question))
                    .getJSONArray("variants")
            for (i in 0 until dataJsonArr.length()) {
                val testJSON = dataJsonArr.getJSONObject(i)
                val availabletest = AvailableTest()
                availabletest.id = testJSON.getInt("id_picture")
                availabletest.name = testJSON.getString("picture")
                println(availabletest.id)
                println(availabletest.name)
                listTest.add(availabletest)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return listTest.toTypedArray()
    }

    fun getPictureNull(id_question: Int): Array<AvailableTest> {
        val listTest = ArrayList<AvailableTest>()
        try {
            val dataJsonArr = JSONObject(getResponse(PICTURE_NULL + id_question))
                    .getJSONArray("variants")
            for (i in 0 until dataJsonArr.length()) {
                val testJSON = dataJsonArr.getJSONObject(i)
                val availabletest = AvailableTest()
                availabletest.id = testJSON.getInt("id_picture")
                println(availabletest.id)
                listTest.add(availabletest)
            }
        } catch (e: Exception) {
            e.printStackTrace()
            val availabletest = AvailableTest()
            availabletest.id = 0
            listTest.add(availabletest)
            return listTest.toTypedArray()
        }
        return listTest.toTypedArray()
    }

    fun checkInternet(): Boolean {
        return getResponse(GET_AVAILABLE_TEST) != ""
    }

    companion object {
        private const val GET_AVAILABLE_TEST = "http://sl-terms.gearhostpreview.com/api/available_test.php"
        private const val GET_ID_QUESTIONS = "http://sl-terms.gearhostpreview.com/api/get_questions.php?id_test="
        private const val GET_TEXT_QUESTION = "http://sl-terms.gearhostpreview.com/api/name_question.php?id_question="
        private const val INSERT_STUDENT = "http://sl-terms.gearhostpreview.com/api/Insert_student.php?surname="
        private const val GET_ID_VARIANT_NAME_VARIANT = "http://sl-terms.gearhostpreview.com/api/get_options.php?id_question="
        private const val ANSWER_TO_CUR_QUESTION = "http://sl-terms.gearhostpreview.com/api/insert_answer.php?id_student="
        private const val NUMBER_OF_CORRECT_ANSWERS = "http://sl-terms.gearhostpreview.com/api/number_of_correct_answers.php?id_student="
        private const val PICTURE = "http://sl-terms.gearhostpreview.com/api/picture.php?id_question="
        private const val PICTURE_NULL = "http://sl-terms.gearhostpreview.com/api/null_picture.php?id_question="
        var dbt: DataBaseTest? = null
    }
}